### DISCUSSION TITLE
After `ts_query_cursor_set_point_range`, can I loop through captures in order and get the same captures as when looping through the matches?

### DISCUSSION THREAD

**Discussion by Danielkonge** [tree-sitter/tree-sitter#3466] [Q&A] [answered]
(Note: The following is built on my use of treesitter in neovim, but from my understanding it is a question more about treesitter itself than anything in neovim. Also, I know my example is a bit vague, but hopefully the question makes sense?)

Assume I want to capture e.g. parentheses in this very simple example where I get a match for every pair `()` (assuming I have set it up so the parentheses on lines 2 and 5 give a match and similarly lines 3 and 4 give match):
```
1: ...
2: (
3: (
4: ) 
5: )
6: ...
```
And assume I set a cursor range covering only lines 2 and 3. 

Then, going through the matches, I get all the parentheses from lines 2 to 5, since the match of `()` from lines 2 and 5 overlap with my given range, but going through the captures I only get the parentheses from line 3 and 4.

Is there any way of keeping the cursor range and still go through all parentheses from lines 2 to 5 in order? (If I go through the matches, the parentheses won't be in order.)

**Comment by amaanq** [comment]
It'd be really helpful if you could at least detail the language, a code snippet, a byte range, and the query you have that's giving these unexpected results. "Parenthesis not being in order with matches" isn't helpful if there's no associated query code to go along with it

**Reply by Danielkonge** [reply]
I will try to a bit more clear. Also just to clarify, I don't think the current behavior is unexpected or wrong, I would just like it if there was some option I could use to get slightly different behavior. (I thought there might already be a way to do what I want, but I am not sure.)

What I want is a way to get around a problem I have when creating rainbow delimiters in neovim (more details here: https://github.com/neovim/neovim/issues/27296).

Say I have the following Lua query
```query
(table_constructor
  "{" @delimiter
  "}" @delimiter @sentinel) @container

(field
  "[" @delimiter
  "]" @delimiter @sentinel) @container

(arguments
  "(" @delimiter
  ")" @delimiter @sentinel) @container
```
and I want to highlight delimiters in the Lua code
```lua
local tmp = {
  [1] = {1},
  [2] = {
    a = print(|)
  },
}
```
where `|` is the cursor position. 

If I go through the whole Lua file with `ts_query_cursor_next_capture`, then I get the captures of all the delimiters in the order I would expect:
```
container -- from outer {}
delimiter -- {   from outer {}

container -- from [1]
delimiter -- [   from [1]
delimiter -- ]   from [1]
sentinel  -- ]   from [1], which ends here

container --     from first inner {}
delimiter -- {   from first inner {}
delimiter -- }   from first inner {}
sentinel  -- }   from first inner {}, which ends here

container -- from [2]
delimiter -- [   from [2]
delimiter -- ]   from [2]
sentinel  -- ]   from [2], which ends here

container --     from second inner {}
delimiter -- {   from second inner {}

container --     from () in print()
delimiter -- (   from ()
delimiter -- )   from ()
sentinel  -- )   from (), which ends here

delimiter -- }   from second inner {}
sentinel  -- }   from second inner {}, which ends here

delimiter -- }   from outer {}
sentinel  -- }   from outer {}, which ends here
```

If I set ts_query_cursor_set_point_range to only cover the cursor line, and then go through the captures, then I get something like:
```
container --      from outer {}

container --      from [2]

container --      from second inner {}

container --     from () in print()
delimiter -- (   from ()
delimiter -- )   from ()
sentinel  -- )   from (), which ends here
```

Here I get the `container`s because their match overlap with the cursor line (from my understanding). Notably I don't get `container`s from `[1]` or the first `{}`.

But I would like an option to get all the captures in any matches overlapping with the cursor line. I.e., in the above I would like to get something like:
```
container -- from outer {}
delimiter -- {   from outer {}

container -- from [2]
delimiter -- [   from [2]
delimiter -- ]   from [2]
sentinel  -- ]   from [2], which ends here

container --     from second inner {}
delimiter -- {   from second inner {}

container --     from () in print()
delimiter -- (   from ()
delimiter -- )   from ()
sentinel  -- )   from (), which ends here

delimiter -- }   from second inner {}
sentinel  -- }   from second inner {}, which ends here

delimiter -- }   from outer {}
sentinel  -- }   from outer {}, which ends here
```

If I loop through the matches instead of captures, I would already get the captures I want, but I wouldn't get them in order like when I loop through the captures, and I would prefer to get them in order.

I don't know if this was clear enough? Otherwise I can try to explain it in a different way.

**Reply by amaanq** [reply, marked as answer]
got it, the code + query was very helpful in understanding your problem, so thank you :slightly_smiling_face:. when a capture is returned with `ts_query_cursor_next_capture` (which is a `TSQueryMatch` + the index of the capture), you can access *all* the captures with the `captures` field of the query match. I don't know how this is done in neovim all too well, but it's doable here, although you might be getting the same overall match several time, but with different capture indices, so you'll have to figure how you want to do this on your own (though it is possible)

**Reply by Danielkonge** [reply]
In neovim I can also access all the captures with the `captures` field of the query match, my problem is just that I can't easily get the captures in the order I want. 

I guess the answer is just, that there is no way to get only the captures I want in order. I can either get more captures (including `[1]` and the first `{}` above) or the captures I want, but not ordered (via the matches). 

Thank you for the help.

**Reply by amaanq** [reply]
ah I see, yeah, there's no way to get that specific order with a narrower range passed in because some of those captures just do not overlap with the desired range, sorry about that
